File: algorithms/code/leetcode/lc070_climbing_stairs/lc070_climbing_stairs.py
#!/usr/bin/env python3
# https://leetcode.com/problems/climbing-stairs/

import unittest


class Solution:
    def __climbStairs(self, n, at, lut):
        if at + 2 == n:
            return 2
        if at + 1 == n:
            return 1
        if lut[at] == 0:
            lut[at] = self.__climbStairs(n, at + 1, lut)
            lut[at] += self.__climbStairs(n, at + 2, lut)
        return lut[at]

    def climbStairs(self, n: int) -> int:
        lut = [0] * n
        return self.__climbStairs(n, 0, lut)


class TestCode(unittest.TestCase):
    def test_20(self):
        self.assertEqual(10946, Solution().climbStairs(20))

File: algorithms/code/leetcode/lc070_climbing_stairs/lc070_climbing_stairs.rs
// https://leetcode.com/problems/climbing-stairs/
// #easy

pub struct Solution;

impl Solution {
    fn climb_stairs_internal(n: i32, at: i32, lut: &mut [i32]) -> i32 {
        if at + 2 == n {
            return 2;
        }
        if at + 1 == n {
            return 1;
        }
        if lut[at as usize] == 0 {
            lut[at as usize] = Self::climb_stairs_internal(n, at + 1, lut)
                + Self::climb_stairs_internal(n, at + 2, lut);
        }
        lut[at as usize]
    }

    pub fn climb_stairs(n: i32) -> i32 {
        let mut lut = vec![0; n as usize];
        Self::climb_stairs_internal(n, 0, &mut lut)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_20() {
        assert_eq!(Solution::climb_stairs(20), 10946);
    }
}
